src/Stratis.Bitcoin/BlockPulling2/BlockPuller.cs

```diff
@@ -0,0 +1,771 @@
+
+            this.networkPeerRequirement = new NetworkPeerRequirement
+            {
+                MinVersion = protocolVersion,
+                RequiredServices = NetworkPeerServices.Network
+            };
+
+            this.cancellationSource = new CancellationTokenSource();
+            this.random = new Random();
+
+            this.OnDownloadedCallback = callback;
+            this.chainState = chainState;
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
+        }
+
+        public void Initialize()
+        {
+            this.logger.LogTrace("()");
+
+            this.assignerLoop = this.AssignerLoopAsync();
+            this.stallingLoop = this.StallingLoopAsync();
+
+            this.logger.LogTrace("(-)");
+        }
+
+        public double GetAverageBlockSizeBytes()
+        {
+            return this.averageBlockSizeBytes.Average;
+        }
+
+        private int GetTotalSpeedOfAllPeersBytesPerSecLocked()
+        {
+            return this.pullerBehaviorsByPeerId.Sum(x => x.Value.SpeedBytesPerSecond);
+        }
+
+        /// <summary>Should be called when a peer claims a new tip.</summary>
+        /// <param name="peer">The peer.</param>
+        /// <param name="newTip">New tip.</param>
+        public void NewPeerTipClaimed(INetworkPeer peer, ChainedHeader newTip)
+        {
+            this.logger.LogTrace("({0}:{1},{2}:'{3}')", nameof(peer.Connection.Id), peer.Connection.Id, nameof(newTip), newTip);
+
+            lock (this.lockObject)
+            {
+                int peerId = peer.Connection.Id;
+
+                if (this.peerIdsToTips.ContainsKey(peerId))
+                {
+                    this.peerIdsToTips.AddOrReplace(peerId, newTip);
+
+                    this.logger.LogDebug("Tip for peer with id {0} was changed to '{1}'.", peerId, newTip);
+                }
+                else
+                {
+                    bool supportsRequirments = this.networkPeerRequirement.Check(peer.PeerVersion);
+
+                    if (supportsRequirments)
+                    {
+                        this.peerIdsToTips.AddOrReplace(peerId, newTip);
+                        this.pullerBehaviorsByPeerId.Add(peerId, peer.Behavior<BlockPullerBehavior>());
+
+                        this.logger.LogDebug("New peer with id {0} was added.", peerId);
+                    }
+                }
+            }
+
+            this.logger.LogTrace("(-)");
+        }
+
+        /// <summary>Should be called when peer is disconnected.</summary>
+        /// <param name="peerId">Unique peer identifier.</param>
+        public void PeerDisconnected(int peerId)
+        {
+            this.logger.LogTrace("({0}:{1})", nameof(peerId), peerId);
+
+            lock (this.lockObject)
+            {
+                this.peerIdsToTips.Remove(peerId);
+                this.pullerBehaviorsByPeerId.Remove(peerId);
+
+                this.ReleaseAssignments(peerId);
```

Comment:
Locked?

Reply:
No, it uses lock inside. 

But it looks like I had a deadlokk there so fixed.